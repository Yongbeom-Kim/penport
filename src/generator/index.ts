export * from "./generator";
export * from "./tailwind_generator";