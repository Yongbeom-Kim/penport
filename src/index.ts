import { program } from "commander";

program
  .command("export")
  .description("Export alls themes from Penpot")
  .action(async () => {
    console.log("Exporting themes...");
  });

program.parse(process.argv);
