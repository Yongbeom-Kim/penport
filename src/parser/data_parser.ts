import { UnrecoverableError } from "../types/errors";
import { Color, ColorSchema, TypographySchema, Typography } from "../types/penpot";

export const parseTypographyData = (data: string): Typography => {
  try {
    const typographyData = JSON.parse(data);
    typographyData.lineHeight = Number.parseFloat(String(typographyData.lineHeight));
    typographyData.letterSpacing = Number.parseFloat(String(typographyData.letterSpacing));
    const typography = TypographySchema.parse(typographyData);
    return typography;
  } catch (error) {
    throw new UnrecoverableError(`Failed to parse typography data: ${error.message}`);
  }
}

export const parseColorData = (data: string): Color => {
  try {
    const colorData = JSON.parse(data);
    const color = ColorSchema.parse(colorData);
    return color;
  } catch (error) {
    throw new UnrecoverableError(`Failed to parse color data: ${error.message}`);
  }
}
