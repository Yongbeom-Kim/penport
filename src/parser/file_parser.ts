import { PathLike, readFileSync } from "node:fs";
import AdmZip from "adm-zip";
import { UnrecoverableError } from "../types/errors";

export const unzipPenpotFile = (file: PathLike | Buffer): AdmZip => {
  const zipFile = file instanceof Buffer ? file : readFileSync(file);
  const zip = new AdmZip(zipFile);
  
  const zipEntries = zip.getEntries();
  if (zipEntries.length === 0) {
    throw new UnrecoverableError("No entries found in the Penpot file.");
  }
  
  return zip;
}

export const isColorEntry = (entry: AdmZip.IZipEntry): boolean => {
  const result = /colors\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\.json$/.test(entry.entryName as string)
  return result;
}

export const isTypographyEntry = (entry: AdmZip.IZipEntry): boolean => {
  const result = /typographies\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\.json$/.test(entry.entryName as string)
  return result;
}



