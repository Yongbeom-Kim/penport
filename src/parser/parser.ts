import { PathLike } from "node:fs";
import { isColorEntry, isTypographyEntry, unzipPenpotFile } from "./file_parser";
import { parseColorData, parseTypographyData } from "./data_parser";
import { Color, Typography } from "../types/penpot";

export const parsePenpotFile = (penpotFile: PathLike | Buffer): { typographies: Typography[], colors: Color[] } => {
  const unzippedPenpotFile = unzipPenpotFile(penpotFile);
  const entries = unzippedPenpotFile.getEntries();
  
  const typographies = entries
    .filter(isTypographyEntry)
    .map(entry => entry.getData().toString())
    .map(data => parseTypographyData(data));
  const colors = entries
    .filter(isColorEntry)
    .map(entry => entry.getData().toString())
    .map(data => parseColorData(data));

  return { typographies, colors };
}