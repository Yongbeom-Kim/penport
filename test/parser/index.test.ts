import { parsePenpotFile } from "../../src/parser";

describe("parse Base Test Case penpot file", () => {
  it("should unzip a penpot file", () => {
    const filePath = "test/resources/Base Test Case.penpot";
    const result = parsePenpotFile(filePath);
    expect(result.typographies.length).not.toBe(0);
    expect(result.colors.length).not.toBe(0);
    expect(result).toMatchSnapshot();
  });
});