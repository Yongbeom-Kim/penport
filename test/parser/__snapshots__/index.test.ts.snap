// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse Base Test Case penpot file should unzip a penpot file 1`] = `
{
  "colors": [
    {
      "color": "#ecfeff",
      "id": "94b88e59-80ef-8032-8006-4a0e596c18b3",
      "modifiedAt": "2025-06-04T14:57:40.598Z",
      "name": "tailwind-cyan-50",
      "opacity": 1,
      "path": "cy",
    },
    {
      "color": "#6366f1",
      "id": "94b88e59-80ef-8032-8006-4a0e6fa7b977",
      "modifiedAt": "2025-06-04T14:57:44.746Z",
      "name": "tailwind-indigo-500 a-50",
      "opacity": 0.5,
      "path": "id",
    },
    {
      "color": "#1e2932",
      "id": "bfce85a9-aaf9-8010-8006-4a20ce4d3060",
      "modifiedAt": "2025-06-04T14:57:56.074Z",
      "name": "#1e2932",
      "opacity": 1,
    },
  ],
  "typographies": [
    {
      "fontFamily": "sourcesanspro",
      "fontId": "sourcesanspro",
      "fontSize": "14",
      "fontStyle": "normal",
      "fontVariantId": "regular",
      "fontWeight": "400",
      "id": "94b88e59-80ef-8032-8006-4a0c890124e0",
      "letterSpacing": 0,
      "lineHeight": 1.2,
      "modifiedAt": "2025-06-04T13:29:39.727Z",
      "name": "source sans pro 14 regular",
      "path": "ssp",
      "textTransform": "none",
    },
    {
      "fontFamily": "sourcesanspro",
      "fontId": "sourcesanspro",
      "fontSize": "18",
      "fontStyle": "italic",
      "fontVariantId": "bolditalic",
      "fontWeight": "bold",
      "id": "94b88e59-80ef-8032-8006-4a0cc0266f99",
      "letterSpacing": 0,
      "lineHeight": 1.2,
      "modifiedAt": "2025-06-04T13:30:24.183Z",
      "name": "Source Sans Pro Bolditalic",
      "path": "ssp",
      "textTransform": "none",
    },
    {
      "fontFamily": "sourcesanspro",
      "fontId": "sourcesanspro",
      "fontSize": "10",
      "fontStyle": "normal",
      "fontVariantId": "200",
      "fontWeight": "200",
      "id": "94b88e59-80ef-8032-8006-4a0cfb7de80a",
      "letterSpacing": 0,
      "lineHeight": 1.2,
      "modifiedAt": "2025-06-04T13:31:25.579Z",
      "name": "Source Sans Pro 200",
      "path": "ssp",
      "textTransform": "none",
    },
    {
      "fontFamily": "Stalemate",
      "fontId": "gfont-stalemate",
      "fontSize": "10",
      "fontStyle": "normal",
      "fontVariantId": "regular",
      "fontWeight": "400",
      "id": "94b88e59-80ef-8032-8006-4a0d3e146026",
      "letterSpacing": 0,
      "lineHeight": 1.2,
      "modifiedAt": "2025-06-04T14:31:56.891Z",
      "name": "Stalemate Regular",
      "path": "st",
      "textTransform": "capitalize",
    },
    {
      "fontFamily": "Stalemate",
      "fontId": "gfont-stalemate",
      "fontSize": "24",
      "fontStyle": "normal",
      "fontVariantId": "regular",
      "fontWeight": "400",
      "id": "94b88e59-80ef-8032-8006-4a0d45d7a24b",
      "letterSpacing": 0,
      "lineHeight": 1.2,
      "modifiedAt": "2025-06-04T13:32:44.669Z",
      "name": "Stalemate Regular 24 line spacing",
      "path": "st",
      "textTransform": "none",
    },
    {
      "fontFamily": "Stalemate",
      "fontId": "gfont-stalemate",
      "fontSize": "43",
      "fontStyle": "normal",
      "fontVariantId": "regular",
      "fontWeight": "400",
      "id": "94b88e59-80ef-8032-8006-4a0d8e4894b9",
      "letterSpacing": 2.4,
      "lineHeight": 3,
      "modifiedAt": "2025-06-04T13:33:56.286Z",
      "name": "Stalemate Regular all caps 24",
      "path": "st",
      "textTransform": "uppercase",
    },
    {
      "fontFamily": "sourcesanspro",
      "fontId": "sourcesanspro",
      "fontSize": "14",
      "fontStyle": "normal",
      "fontVariantId": "regular",
      "fontWeight": "400",
      "id": "bfce85a9-aaf9-8010-8006-4a20d64efe6d",
      "letterSpacing": 0,
      "lineHeight": 1.2,
      "modifiedAt": "2025-06-04T14:58:14.225Z",
      "name": "no-group",
      "path": "",
      "textTransform": "none",
    },
  ],
}
`;
